### PR TITLE
feat: Add ddev config --update flag, fixes #6035

### LINKED ISSUE #6035: Improve support for the new `drupal`project type
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

While testing around with https://github.com/justafish/ddev-drupal-core-dev i've identified two scenarios where the newly added project type `drupal` might be improved further: 

1. If you take a look at the current recommended quickstart guides using the `drupal` project type: https://ddev.readthedocs.io/en/latest/users/quickstart/#drupal . you run `ddev config` on an empty directory there. the logics behind that command has no codebase at hand to determine the correct version of drupal that way. the codebase is pulled two steps after the first step on `ddev composer create`. for that reason ddev is falling back to the default values, php 8.2 and corepack enable false. but if you want to install a drupal 11 instance that requires PHP 8.3 and corepack enabled for example. 

2. The other case is if you have an existing project with the drupal10 project type and after upgrading from drupal 1.22.7 to 1.23.0 and you change the project type on `ddev config` from `drupal10` to `drupal` the settings remain the same. so in case you were on a drupal 11 instance before upgrading the newly added corepack_enable will stay set to false.

### Describe your solution

For point one. the config might be reevaluated and updated after `ddev composer create` is run and the codebase is in place. @rfay already ideated to introduce an `--update` flag to `ddev config` as well. i am not sure if i would make something that is altering the settings completely free accessibly (might entail unwanted sideeffects). but triggering such a functionality right after `ddev composer create` is run might be safe and fix the problem. but on the other hand a freely available command `ddev config --update` would also fix the second point. Not sure just wanted to open a discussion. 

### Describe alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by penyaskito**
## The Issue

* #6035 

## How This PR Solves The Issue

Allow overriding after detecting actual project type

## TODO:

- [x] Add a test for --update
- [x] Consider making --update mutually exclusive with other options
- [x] Note to people after `ddev composer create` that they may want to do it.
- [x] drupal8 should not override existing database
- [x] Docs need to explain/expose `ddev config --auto` and `ddev config --update`.  Difference, when to use, etc. How they relate to `ddev config`.
- [ ] Consider changing quickstarts to just use `ddev config --auto` for initial configuration

## Manual Testing Instructions

Assuming you have an existing D9 project, and updating to Drupal 10, you can run 

```console
ddev config --update
```

.ddev/config.yaml changed as expected:

```diff
 name: pantheon-test
-type: drupal9
+type: drupal
 docroot: web
-php_version: "8.1"
+php_version: "8.3"
```

## Related Issue Link(s)

Alternative PR to 
* https://github.com/ddev/ddev/pull/6053
* https://github.com/ddev/ddev/pull/5887

### CODE REVIEW

**Review by rfay — CHANGES_REQUESTED** [review]
This will need a test of course!

**Comment by rfay** [comment]
Related:
* https://github.com/ddev/ddev/pull/6058

**Comment by rpkoller** [comment]
i've tested the artifact with all the basic quickstarts from d8-d11. for d10 and d11 it works just fine the other two (8 & 9) probably just need drupal8 and drupal9 as their projecttype cuz those problems arent solvable with the `ddev config --update`command. 

but flag for config functions well. just one thought i already ran by @rfay, would it somehow be possible to trigger a prompt after `ddev composer create` was run telling the user something like it looks like the current config settings are not in line with your current code base do you want to run `config --update` . i think that update command will work fine for folks working through a quickstart but i am worried (looking at myself for example) that from time to time one will forget about running the config update command after a composer create on a project with the drupal app type. so having a reminder and failsafe in place mit be a thoughtful step?

**Comment by rfay** [comment]
TODO:
- [ ] Autodetection is not working correctly. For example `ddev config --auto` will not detect a project of type `drupal`. Perhaps it's not doing the right thing with docroot? You can test by checking out drupal core and `ddev config --auto`
- [ ] Needs a test for this new behavior, or two. At least one in pkg/ddevapp, perhaps one in cmd.
- [ ] It needs to not change the database type
- [ ] Open question: Should it change the docroot if a legit docroot is found that is not the configured one?

**Comment by rfay** [comment]
For least confusion you'll want to rebase right away to get the new config overrides that went in in
* https://github.com/ddev/ddev/pull/6058

**Comment by rfay** [comment]
Just a note: `ddev config --auto` is actually just `ddev config` with a dummy argument. It doesn't do anything, and the result should be the same as typing `ddev config` and accepting the defaults.

**Comment by rpkoller** [comment]
> Just a note: `ddev config --auto` is actually just `ddev config` with a dummy argument. It doesn't do anything, and the result should be the same as typing `ddev config` and accepting the defaults.

i can confirm, with the latest patch i've cloned the drupal repo in an empty directory with
`git clone https://git.drupalcode.org/project/drupal.git .`

with `ddev config --auto --omit-containers=db --disable-settings-management`i then get a drupal project with php 8.3 and corepack_enable true 

while with `ddev config --omit-containers=db --disable-settings-management` i get a php project with  php 8.2 and corepack_enable false

**Comment by rpkoller** [comment]
I think i found another edge case. I followed the following steps: 

1. create an empty directory. 
2. run `ddev config`, use the default for the name, use `docroot`for docroot, use`drupal`for the project type.
3.  then you realize shoot you have used `docroot` instead going with `web`, so you run `ddev config`another time,  for name and project stick with the default name and drupal for docroot you use web now.  
4. `ddev composer create drupal/recommended-project:^9.5`

you get the following error 

```
Failed to create project: '/Users/rkoller/Sites/dd9/docroot' is not allowed to be present. composer create needs to be run on a clean/empty project with only the following paths: [web web/sites web/sites/default web/sites/default/.gitignore web/sites/default/files web/sites/default/files/sync web/sites/default/settings.ddev.php web/sites/default/settings.php] - please clean up the project before using 'ddev composer create'
```
looks like if you update the docroot with the manual steps in `ddev config` or if you run a `ddev config --update --docroot="web" --project-type="drupal"` after setting docroot to docroot with the initial manual step, you end up with two folders, docroot and web. would it make sense to delete the previous folder you are upgrading from or at least leave a note after running the second upgrade changing the docroot to prevent that you run into the "failed" error?

i see the potential problem if someone at a later point changes the docroot accidentaly on a fully set up site, that would wipe out most of its assets.

File: cmd/ddev/cmd/config.go [hunk 1]
```diff
@@ -438,7 +445,7 @@ func handleMainConfigArgs(cmd *cobra.Command, _ []string, app *ddevapp.DdevApp)
 
 	// App overrides are done after app type is detected, but
 	// before user-defined flags are set.
-	err = app.ConfigFileOverrideAction()
+	err = app.ConfigFileOverrideAction(updateArg || autoArg)
```
[the remark this answers is not in the document]
Comment (rfay) [reply cmd/ddev/cmd/config.go:448]: I suspect what we have to do is make `ddev config --update` incompatible with all other flags. In other words, you can't specify anything else, and when you do it it updates everything but database (unless database doesn't already exist?)

File: pkg/ddevapp/craftcms.go [hunk 2]
```diff
@@ -150,7 +150,19 @@ func craftCmsPostStartAction(app *DdevApp) error {
...
+		return err
+	}
+	recommendedDBType := "mysql:8.0"
+	if dbType == "" {
+		// Assume that we don't have a database yet
+		app.Database = DatabaseDesc{nodeps.MySQL, nodeps.MySQL80}
+		util.Success("Configuring Craft CMS project with database type '%s'", recommendedDBType)
+	} else if dbType != recommendedDBType {
+		util.Warning("Craft CMS project already has database type set to non-default: %s", dbType)
+	}
```
[the remark this answers is not in the document]
Comment (penyaskito) [reply pkg/ddevapp/craftcms.go:165]: https://github.com/ddev/ddev/pull/6055/files#diff-bd1e2b41cfb1bf9898d43f3203d49e9c2e9af1163977e9af6a5472cffaa9416cR436

We could check there if previousApp.dbType was set and different than the new app.dbType (the new recommended one), revert the change and fire the warning instead.

[the remark this answers is not in the document]
Comment (penyaskito) [reply pkg/ddevapp/craftcms.go:165]: Some "pseudocode":

```
        currentDb := app.Database  
		err = app.ConfigFileOverrideAction(true)
		if err != nil {
			util.Warning("ConfigOverrideAction failed: %v")
		}
        if currentDb != app.Database { // Whatever this means, as we would need to check dbtype + version and it's a struct 
    		util.Warning("Project already has database type set to non-default: %s, if you want to upgrade it follow x steps", dbType)
            // Restore it
			app.Database = currentDb
			app.writeConfig();
       }
```

[the remark this answers is not in the document]
Comment (rfay) [reply pkg/ddevapp/craftcms.go]: Thanks, that's a far better technique. Added https://github.com/ddev/ddev/pull/6055/commits/656b32462fcad014d4def13ba604feca45fdd9eb with that approach, will appreciate your review.

[the remark this answers is not in the document]
Comment (penyaskito) [reply pkg/ddevapp/craftcms.go]: I was suggesting a different place, but actually prefer what you did. Looks great to me!

File: pkg/ddevapp/apptypes_test.go [hunk 3]
```diff
@@ -106,11 +106,59 @@ func TestConfigOverrideAction(t *testing.T) {
 		app.PHPVersion = newVersion
 		err = app.WriteConfig()
 		assert.NoError(err)
-		err = app.ConfigFileOverrideAction()
+		err = app.ConfigFileOverrideAction(false)
 		assert.NoError(err)
 		// But with a config that has been written with a specified version, the version should be untouched by
```
Comment (penyaskito) [inline pkg/ddevapp/apptypes_test.go:111]:
```suggestion
		// But with a config that has been written with a specified version, the version should be untouched 
```

File: pkg/ddevapp/apptypes_test.go [hunk 4]
```diff
@@ -106,11 +106,59 @@ func TestConfigOverrideAction(t *testing.T) {
...
+func TestConfigOverrideActionOnExistingConfig(t *testing.T) {
+	assert := asrt.New(t)
+	origDir, _ := os.Getwd()
+
+	// This will only work for those project types defining configOverrideAction and altering php version
+	appTypes := map[string]string{
+		nodeps.AppTypeCakePHP: nodeps.PHP83,
+		nodeps.AppTypeDrupal6: nodeps.PHP56,
+		nodeps.AppTypeDrupal7: nodeps.PHP82,
+		// For AppTypeDrupal we can't guess a version without a working installation.
```
Comment (penyaskito) [inline pkg/ddevapp/apptypes_test.go:127]: Now that you added e.g. `cmd/ddev/cmd/testdata/TestConfigUpdate/drupal11-composer/` we could improve test coverage here.

**Review by penyaskito — COMMENTED** [review]
Didn't manually test it.
But reading it looks pretty solid to me. Left some suggestions, nothing blocking.

File: docs/content/users/quickstart.md [hunk 5]
```diff
@@ -181,11 +182,11 @@ For all versions of Drupal 8+ the Composer techniques work. The settings configu
...
 
     ```bash
-    mkdir my-drupal-site
-    cd my-drupal-site
-    ddev config --project-type=drupal --php-version=8.3 --docroot=web --corepack-enable
+    mkdir my-drupal-site && cd my-drupal-site
+    ddev config --project-type=drupal --php-version=8.3 --docroot=web
     ddev start
     ddev composer create drupal/recommended-project:^11.x-dev
+    ddev config --update
```
[the remark this answers is not in the document]
Comment (rfay) [reply docs/content/users/quickstart.md:189]: We could change the `ddev config` to be just `ddev config --php-version=8.3` and that's all, then `ddev config --update` would finish the job. That's kind of the thought. @rpkoller was trying to make it so the config lines were smaller and less error-prone.

File: pkg/ddevapp/apptypes_test.go (hunk 3 above)
Comment (rfay) [reply pkg/ddevapp/apptypes_test.go:111]: This is intended to read "the version should be untouched by app.ConfigFileOverrideAction()" so it's correct I think.

File: pkg/ddevapp/apptypes_test.go (hunk 4 above)
Comment (rfay) [reply pkg/ddevapp/apptypes_test.go:127]: By a lot. Maybe we should throw this test away and move everything over there.

File: pkg/ddevapp/apptypes_test.go (hunk 3 above)
Comment (penyaskito) [reply pkg/ddevapp/apptypes_test.go:111]: Yes, but the ` app.ConfigFileOverrideAction()` line was removed? Might be a glitch on github diff rendering, didn't look at this locally.

Comment (rfay) [reply pkg/ddevapp/apptypes_test.go:111]:
```suggestion
		// But with a config that has been written with a specified version, the version should be untouched by
		// app.ConfigFileOverrideAction()
```

**Review by rfay — APPROVED** [review]
I think it's ready to go. We may want to follow up with quickstart changes, simplifying the initial `ddev config` to just `ddev config --php-version` for example.

**Review by stasadev — APPROVED** [review]
Tested it with `craftcms` quickstart, have no objections, everything works great, great job!

**Pull request merged**